contracts/modules/capital/CowSwapOperator.sol

```diff
@@ -0,0 +1,240 @@
+
+    // Store the order UID
+    currentOrderUID = orderUID;
+
+    // Sign the Cow order
+    cowSettlement.setPreSignature(orderUID, true);
+
+    // Emit an event
+    emit OrderPlaced(order);
+  }
+
+  function closeOrder(GPv2Order.Data calldata order) external {
+    // Validate there is an order in place
+    require(currentOrderUID.length > 0, "SwapOp: No order in place");
+
+    // Before validTo, only controller can call this. After it, everyone can call
+    if (block.timestamp <= order.validTo) {
+      require(msg.sender == swapController, "SwapOp: only controller can execute");
+    }
+
+    validateUID(order, currentOrderUID);
+
+    // Check how much of the order was filled, and if it was fully filled
+    uint256 filledAmount = cowSettlement.filledAmount(currentOrderUID);
+    bool fullyFilled = filledAmount == order.sellAmount;
+
+    // Cancel signature and unapprove tokens
+    if (!fullyFilled) {
+      cowSettlement.setPreSignature(currentOrderUID, false);
+      approveVaultRelayer(order.sellToken, 0);
+    }
+
+    // Clear the current order
+    delete currentOrderUID;
+
+    // Withdraw both buyToken and sellToken
+    returnAssetToPool(order.buyToken);
+    returnAssetToPool(order.sellToken);
+
+    // Set swapValue on pool to 0
+    _pool().setSwapValue(0);
+
+    // Emit event
+    emit OrderClosed(order, filledAmount);
+  }
+
+  function returnAssetToPool(IERC20 asset) internal {
+    uint256 balance = asset.balanceOf(address(this));
+
+    if (balance == 0) {
+      return;
+    }
+
+    if (address(asset) == address(weth)) {
+      weth.withdraw(balance); // Unwrap WETH
+      payable(address(_pool())).transfer(balance); // Transfer ETH to pool
+    } else {
+      asset.transfer(address(_pool()), balance); // Transfer ERC20 to pool
+    }
+  }
+
+  function isSellingEth(GPv2Order.Data calldata order) internal view returns (bool) {
+    return address(order.sellToken) == address(weth);
+  }
+
+  function isBuyingEth(GPv2Order.Data calldata order) internal view returns (bool) {
+    return address(order.buyToken) == address(weth);
+  }
+
+  function validateBasicCowParams(GPv2Order.Data calldata order) internal view {
+    require(order.sellTokenBalance == GPv2Order.BALANCE_ERC20, "SwapOp: Only erc20 supported for sellTokenBalance");
+    require(order.buyTokenBalance == GPv2Order.BALANCE_ERC20, "SwapOp: Only erc20 supported for buyTokenBalance");
+    require(order.receiver == address(this), "SwapOp: Receiver must be this contract");
+    require(
+      order.validTo >= block.timestamp + MIN_VALID_TO_PERIOD,
+      "SwapOp: validTo must be at least 10 minutes in the future"
+    );
+    require(
+      order.validTo <= block.timestamp + MAX_VALID_TO_PERIOD,
+      "SwapOp: validTo must be at most 60 minutes in the future"
+    );
+  }
+
+  function approveVaultRelayer(IERC20 token, uint256 amount) internal {
+    token.approve(cowVaultRelayer, amount); // infinite approval
```

Comment:
1. The comment is misleading
2. The return value of the approve function is not checked and depending on the token it may be missing. Use OZ's safeApprove from SafeERC20 because it checks the optionally returned boolean. Beware that a side effect of the safeApprove method is that it checks the allowance to be zero and reverts otherwise (unwanted side effect but I think it's fine for our scenario because we reset the allowance anyway).

Reply:
I just checked the safeApprove function and it seems it's deprecated in favor of safeIncreaseAllowance (https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/utils/SafeERC20.sol#L45), althought the behavior is not the same. I favor the latter as well since if for any reason the allowance is not 0, the contract will be usable, What do you think?